Add support for `AdminRequest` Sessions

amcalgates opened the pull request:
<!-- 🎉 Thank you for submitting a pull request! 🎉  -->

## Summary
Adds support for `AdminRequest` Begin/End session as seen in [Create a session documentation](https://docs.adyen.com/point-of-sale/shopper-engagement/create-session/).

All added classes/additions mirror the naming found in that documentation, however I had to introduce a new class which I named `SessionContainer` to combine the `Session` and `[Operation]` objects for JSON serialization, which can then be base64 encoded to a string and used for `AdminRequest.ServiceIdentification`

Please let me know if you'd prefer a different name for `SessionContainer` - I couldn't find a similar concept in adyen-terminal-api-ios to base the naming off of.

## Tested scenarios
I've tested this with both Begin and End session calls using an AMS1 device.

amcalgates committed:
add support for AdminRequest Begin/End Session

amcalgates committed:
tweak SessionContainer init

kaphacius commented:
Hi @amcalgates,

Thank you so much for taking the time to contribute to our repository.
We will check your request and get back to you soon.

Regards,
Yurii
Adyen

kaphacius commented on Sources/TerminalAPIKit/Models/Operation.swift:
```diff
@@ -0,0 +1,32 @@
+public final class Operation: Codable {
+
+    /// the type of request after which the waiting screen shows: InputRequest, DisplayRequest, and Payment
+    public let type:OperationType
```
can we have a space between the variable name and the type? (applies to multiple places in the PR)

kaphacius commented on Sources/TerminalAPIKit/Models/Operation.swift:
```diff
@@ -0,0 +1,32 @@
+    public let type:OperationType
+    
+    /// For now, the Operation.Variant can only have WaitingScreen as value. This shows the waiting screen animation on the terminal after every request until you end the session.
+    public let variant:String
```
this probably would be better an as enum (with one case :) )

kaphacius commented on Sources/TerminalAPIKit/Models/Operation.swift:
```diff
@@ -0,0 +1,32 @@
+    /// For now, the Operation.Variant can only have WaitingScreen as value. This shows the waiting screen animation on the terminal after every request until you end the session.
+    public let variant:String
+    
+    /// UserInterfaceData.WaitingScreenTimeoutMs: how long the waiting screen is shown on the terminal display, in milliseconds.
```
we probably don't need to mention the field name, just an explanation is enough, e.g.
```swift
     /// UserInterfaceData.WaitingScreenTimeoutMs: how long the waiting screen is shown on the terminal display, in milliseconds.
```

amcalgates replied:
Your example comment is the same as my comment - can you clarify what you want here? Currently `UserInterfaceData` only has one property (and it's required): `waitingScreenTimeoutMs`. Perhaps this property just doesn't have a comment?

amcalgates replied:
For reference, I just copy comments from the Terminal API docs: https://docs.adyen.com/point-of-sale/shopper-engagement/create-session/

kaphacius replied:
apologies for this. seems like i forgot to write my own suggestion for the comment. what i meant was that we should not mention the names of the fields in the comments, since they don't match the actual swift field names.
We can simply mention that `userInterfaceData` contains UI parameters related to the operation.

kaphacius commented on Sources/TerminalAPIKit/Models/Operation.swift:
```diff
@@ -0,0 +1,32 @@
+    /// UserInterfaceData.WaitingScreenTimeoutMs: how long the waiting screen is shown on the terminal display, in milliseconds.
+    public let userInterfaceData:UserInterfaceData
+    
+    public init(type: OperationType, variant: String, userInterfaceData: UserInterfaceData) {
```
a docc-style comment for the initializer would be nice here (applies to multiple places in the PR)

kaphacius commented on Sources/TerminalAPIKit/Models/Operation.swift:
```diff
@@ -0,0 +1,32 @@
+
+public final class Operation: Codable {
+
+    /// the type of request after which the waiting screen shows: InputRequest, DisplayRequest, and Payment
```
the docc comment should be capitalized (applies to multiple places in the PR)

kaphacius commented on Sources/TerminalAPIKit/Models/Session.swift:
```diff
@@ -0,0 +1,27 @@
+    /// the unique reference of the session.
+    public let id: Int
+    
+    /// set the value to Begin to start the session, set the value to End to end the session.
```
maybe we can rewrite the comment here as folows
```swift
The type of the transaction. Use ``SessionType/Begin`` to start a session, and ``SessionType/End`` to end the session.
```

kaphacius commented on Sources/TerminalAPIKit/Models/SessionContainer.swift:
```diff
@@ -0,0 +1,25 @@
+
+public final class SessionContainer: Codable {
+    public let session:Session
+    public let operation:[Operation]?
```
~~looks like, according to the [docs](https://docs.adyen.com/point-of-sale/shopper-engagement/create-session/#start-session), the `operation` is mandatory, so the type of it should probably be non-optional~~

this is wrong, please disregard

amcalgates committed:
Per Yurii@Adyen:
- Add `OperationVariant` enum
- Add comment stubs for inits to match rest of code base
- Fix spacing in variable declarations
- Capitalize first word in comments

amcalgates commented:
I've made all requested changes, except for one comment I was unsure on (see my inline replies).

kaphacius commented on Sources/TerminalAPIKit/Models/Session.swift:
```diff
@@ -0,0 +1,31 @@
+    
+    /// Initializes the Session.
+    ///
+    /// - Parameter id: Undocumented.
```
can we write also document the parameters here? applies in multiple places in the PR

amcalgates replied:
Latest commit should handle these

kaphacius commented on Sources/TerminalAPIKit/Models/Operation.swift:
```diff
@@ -0,0 +1,37 @@
+
+public final class Operation: Codable {
+
+    /// The type of request after which the waiting screen shows: InputRequest, DisplayRequest, and Payment
```
we probably don't need to mention the options here, since the type is now an `enum`, and the possible values will be restricted to that.

amcalgates replied:
Latest commit should handle this

kaphacius commented on Sources/TerminalAPIKit/Models/Operation.swift:
```diff
@@ -0,0 +1,37 @@
+    public let type: OperationType
+    
+    /// For now, the Operation.Variant can only have WaitingScreen as value. This shows the waiting screen animation on the terminal after every request until you end the session.
+    public let variant: OperationVariant
```
we probably don't need to mention the options here, since the type is now an enum, and the possible values will be restricted to that. given that the `OperationVariant` enum contains only one case, it should be logical that it only supports one variant.

amcalgates replied:
[Adyen's docs](https://docs.adyen.com/point-of-sale/shopper-engagement/create-session/#start-session) do not go into any detail on this, and your comment here doesn't explicitly say what you want. Can you tell me explicitly what you want here for this comment? Otherwise I'll just be guessing, and we will just keep going back and forth.

amcalgates replied:
Latest commit should handle

kaphacius commented on Sources/TerminalAPIKit/Models/Enumerations/SessionType.swift:
```diff
@@ -0,0 +1,13 @@
+import Foundation
+
+public enum SessionType: String, Codable {
+    case begin = "Begin"
```
can we add comments to the type, and every case?
this applies to multiple places in the PR

amcalgates replied:
I searched 125 results of `String, CodingKey` and 54 results of `String, Codable` and none of them have comments. What are you looking for here, since I have no examples to go off of, and I'll be breaking the trend with the rest of the code base?

kaphacius reviewed (commented):
Thanks for updating the PR. I made a few small comments.
Could you also make sure the headers in every file have the following format
```swift
//
//  {FILENAME}.swift
//  TerminalAPIKit
//
//  Copyright (c) 2023 Adyen N.V.
//
```

amcalgates committed:
update comments per @kaphacius

amcalgates commented:
I've made most of the requested changes, and left comments in areas I need more clarification on.

amcalgates commented:
@kaphacius - any follow-up on my comments?

kaphacius commented:
hey @amcalgates
sorry for not getting back on this, things were kind of busy.
i'll approve and merge your MR. thanks for contributing!
